stan-dev/math
# Feature/issue 778 add mpi build

#### Submission Checklist


#### Summary:
Add build for MPI to makefiles

#### Intended Effect:
Trigger automatic build of boost MPI for tests with _mpi_ in their path IF STAN_MPI is defined in the make/local file.

#### How to Verify:
1. Define STAN_MPI in make/local (the value used does not matter).
2. `touch test/unit/fake_mpi_test.cpp`
3. `make test/unit/fake_mpi_test`

This should trigger the automatic build of the boost MPI and its dependencies as dynamic libraries.

#### Side Effects:

#### Documentation:

What we need is a working dynamic library libboost_mpi.dylib and libboost_serialization.dylib (with an .so ending on Linux!). In order to get there we use the boost build system which should be good at figuring out how the MPI installation on a given system needs to be used. If the defaults fail, then the user may himself configure the boost MPI build process by editing the lib/boost/user-config.jam file; the instructions how to do that are found in the boost MPI documentation.

Note that for the matter of the makefiles I am using as target lib/boost/stage/lib/mpi.so, because this files is always generated with the same file ending - on Mac and on Linux.

As I only got MPI working with dynamic linking, there are additional steps on Mac to ensure that the final binary can be executed without additional user interaction. On Mac the absolute paths to the dynamic libraries have to be hardcoded into the final executable. For this to work, the dynamic libraries need to be prepared using then install_name tool. Once that is done, the rpath linker option is sufficient to make things work. On Linux this additional step is not needed.

The plan for cmdstan is to add an if statement to the phony build target. In this if we can then trigger the MPI build whenever STAN_MPI is defined in make/local.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
